library/src/main/java/com/novoda/downloadmanager/DownloadManager.java

```diff
@@ -224,8 +224,16 @@ public File getDownloadsDir() {
     }
 
     @Override
-    public void addCompletedBatch(CompletedDownloadBatch completedDownloadBatch) {
-        downloader.addCompletedBatch(completedDownloadBatch);
+    public void addCompletedBatch(CompletedDownloadBatch completedDownloadBatch) throws IllegalArgumentException {
+        if (alreadyContainsBatch(completedDownloadBatch)) {
+            throw new IllegalArgumentException("CompletedDownloadBatch with id: " + completedDownloadBatch.downloadBatchId() + " already exists.");
+        }
+
+        downloader.addCompletedBatch(completedDownloadBatch, downloadBatchMap);
+    }
+
+    private boolean alreadyContainsBatch(CompletedDownloadBatch completedDownloadBatch) {
+        return downloadBatchMap.get(completedDownloadBatch.downloadBatchId()) != null;
```

Comment:
This is the same as `downloadBatchMap.containsKey(completedDownloadBatch.downloadBatchId())`

Reply:
Very true 😂 memory failed me then 👅